src/pybind/mgr/balancer/module.py

```diff
@@ -754,13 +754,13 @@ def do_crush_compat(self, plan):
 
         # Make sure roots don't overlap their devices.  If so, we
         # can't proceed.
-        roots = pe.target_by_root.keys()
+        roots = list(pe.target_by_root.keys())
         self.log.debug('roots %s', roots)
         visited = {}
         overlap = {}
         root_ids = {}
         for root, wm in six.iteritems(pe.target_by_root):
-            for osd in wm.iterkeys():
+            for osd in list(wm.keys()):
```

Comment:
Should this be 'inter(wm.keys())'?

Reply:
Ah. This would then not run under py2... So what you have here is py2/py3 compliant, right?

Reply:
Modified to use iter(keys)
It runs safe:

a.py:

a = { '1': '2' , '2' : '3' }

XK = iter(a.keys())
XI = iter(a.items())

for x in XK:
    print(x)

for arg, val in XI:
    print("{} {} {} {}".format(type(arg), arg, type(val), val))



PYTHON2
1
2
<type 'str'> 1 <type 'str'> 2
<type 'str'> 2 <type 'str'> 3
PYTHON3
1
2
<class 'str'> 1 <class 'str'> 2
<class 'str'> 2 <class 'str'> 3